src/Nix/Builtins.hs

```diff
@@ -1435,6 +1435,37 @@ hashFileNix nsAlgo nvfilepath = Prim $ hash =<< fileContent
   fileContent = mkNixStringWithoutContext <$> Nix.Render.readFile nvfilepath
 
 
+-- | groupByNix
+-- Groups elements of list together by the string returned from the function f called on 
+-- each element. It returns an attribute set where each attribute value contains the 
+-- elements of list that are mapped to the same corresponding attribute name returned by f.
+groupByNix
+  :: forall e t f m
+   . MonadNix e t f m
+  => NValue t f m
+  -> NValue t f m
+  -> m (NValue t f m)
+groupByNix nvfun nvlist = do
+  list   <- demand nvlist
+  fun    <- demand nvfun
+  (f, l) <- extractP (fun, list)
+  mkNVSet mempty
+    .   fmap (mkNVList . reverse)
```

Comment:
> `reverse` sad face.
> 
> Currently _idk_ how to reduce it here.

We can `flip (<>)` instead, either way is ugly (

Reply:
Dig a little.
* The `reverse` should be fine, is O(n), done in recursion with accumulator, [here](https://hackage.haskell.org/package/base-4.14.3.0/docs/src/GHC-List.html#reverse).
* The `flip (<>)` would be worse, would result in `acc <> one x`, append to the end every encounter, O(n^2). 

as for #880, I believe it could not be get rid of list since `fromListWith :: (v -> v -> v) -> [(k, v)] -> HashMap k v`